src/LayerPlan.cpp

```diff
@@ -1387,12 +1391,21 @@ void LayerPlan::writeGCode(GCodeExport& gcode)
                 gcode.writeTemperatureCommand(prev_extruder, prev_extruder_temp, wait);
             }
         }
-        else if (extruder_plan_idx == 0 && layer_nr != 0 && Application::getInstance().current_slice->scene.extruders[extruder_nr].settings.get<bool>("retract_at_layer_change"))
+        else if (extruder_plan_idx == 0)
         {
-            // only do the retract if the paths are not spiralized
-            if (!mesh_group_settings.get<bool>("magic_spiralize"))
+            const WipeScriptConfig& wipe_config = storage.wipe_config_per_extruder[extruder_plan.extruder_nr];
+            if (wipe_config.clean_between_layers && gcode.getExtrudedVolumeAfterLastWipe(extruder_nr) > wipe_config.max_extrusion_mm3)
             {
-                gcode.writeRetraction(retraction_config);
+                gcode.insertWipeScript(wipe_config);
+                gcode.ResetLastEValueAfterWipe(extruder_nr);
+            }
+            else if (layer_nr != 0 && Application::getInstance().current_slice->scene.extruders[extruder_nr].settings.get<bool>("retract_at_layer_change"))
```

Comment:
This of course makes `retract_at_layer_change` incompatible with `clean_between_layers`, which is fine, but I don't see any handling of this in the frontend (see my comment there).

Reply:
The following condition will not always true even if `clean_between_layers` was turned on:
`if (wipe_config.clean_between_layers && gcode.getExtrudedVolumeAfterLastWipe(extruder_nr) > wipe_config.max_extrusion_mm3)`

I mean, if clean_between_layers is ON, it will not necessarily be working at every layer changing. For those layers when clean between layers was not activated, because last extruded volume value has not exceeded the max_extrusion_mm3 yet, `retract_at_layer_change` may work (it it was also enabled).

Does this logic make sense to you?

Reply:
@vgribinchuk I think what @rburema means is the `else if` part:
```
if (clean_between_layers and volume > max_mm3)
{
    ...
}
<<else>> if (retract_at_layer_change)
{
    ...
}
```
The `else if` here makes `clean_between_layers` and `retract_at_layer_change` exclusive to each other. If this is not intentional, we should remove the `else`.

Reply:
Hi @Lipu,
Yes that is what I meant, but I also understand what @vgribinchuk is saying here:
Namely that it is mut.exclusive _only_ per layer, and that a model _as a whole_ still may be effected by both settings, which means disabling one if the other is on in the front-end will limit possibilities.
I can see that point, but think it should be documented somewhere (something like 'this will influence retract at layer change') where the user can see it.
cheers, Remco.

Reply:
Hi @rburema, @LipuFei ! 
Thank you, gentlemen, for clarifications!
So, I see two options here I can implement.
1. remove `<<else>>` making `clean_between_layers` and `retract_at_layer_change` not mutually exclusive

But this might be a bit confusing, at least from my point of view. Because wipe script provides its own settings for retraction, which could lead to two consecutive retractions at layer change.

So, since wipe script has own settings for retraction, those settings could be applied for retraction at layers where script would be working. On other layers, where script would not be working, `retract_at_layer_change` could be applied. This is the the second option:
2. leave logic as it is, but change description for `clean_between_layers` in front end, like this:
```
"clean_between_layers": {
"description": "Whether to include nozzle wipe G-Code between layers. Enabling this setting could influence behavior of retract at layer change. Please use Wipe Retraction settings to control retraction at layers where the wipe script will be working."
...
```

Let me please know which option works best for you. Or if you see other more appropriate solution.

Reply:
Hi @vgribinchuk,
I think we should leave the logic as it is and change the description, which seems like the option least likely to casue a mess later on.
cheers, Remco.

Reply:
OK, got it. Thank you, @rburema !

Reply:
Description was changed. Please see commit:
https://github.com/Ultimaker/Cura/pull/5356/commits/6e955f780b115a86197b1c8b2e2d892efbfe8cf6